### PR TITLE
add Message class in base.proto

### LINKED ISSUE #1641: Add abstract `Message` in `base.proto`
**Is your feature request related to a problem? Please describe.**

The spec `base.proto` is misaligned from class `aea.mail.base.Message` requirements. 

Specifically, the class `Message` has the field dialogue reference, the message id and the target fields. However, they are repeated in the definition of each `*.proto`, whereas those particular fields could be defined just once in the `base.proto`.

**Describe the solution you'd like**
Define a `Message` Protobuf message, and make all the messages of existing protocols to inherit from it.
For backward compatibility, it should still be possible to just send bytes in an `Envelope`.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Refactor Protobuf files so to avoid code duplication.

## Fixes

Fixes #1641 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/connections/stub/connection.py
Comment (marcofavorito) [inline aea/connections/stub/connection.py:79]: For `test_echo`, This line added a new byte `\xc2` at position 1:

```python
b"\x12\x87\x01\x08\x01\x12@aa88dbbc4ab9ad89aff8682b4474a241055279e40f628aa263a8cd3848672eac*A\n4type.googleapis.com/fetch.aea.Default.DefaultMessage\x12\t*\x07\n\x05hello"

b"\x12\xc2\x87\x01\x08\x01\x12@aa88dbbc4ab9ad89aff8682b4474a241055279e40f628aa263a8cd3848672eac*A\n4type.googleapis.com/fetch.aea.Default.DefaultMessage\x12\t*\x07\n\x05hello"
```

The first is the byte encoding of a default message; the second is the message read by the stub connection before the change.

Comment (DavidMinarsch) [reply aea/connections/stub/connection.py:79]: Not sure what's going on here. Looks like we finally need to address https://github.com/fetchai/agents-aea/issues/1413

File: aea/test_tools/generic.py
Comment (DavidMinarsch) [inline aea/test_tools/generic.py:59]: See above; although we could leave this if we are sure it is always 3 lines for default protocol..hacky I know

File: docs/agent-vs-aea.md
Comment (DavidMinarsch) [inline docs/agent-vs-aea.md:120]: So our new format is much less efficient...

File: aea/mail/base.proto
Comment (marcofavorito) [reply aea/mail/base.proto:3]: yes, sounds good!

File: docs/agent-vs-aea.md
Comment (marcofavorito) [reply docs/agent-vs-aea.md:120]: and it doesn't even contain the full dialogue references!

**Comment by marcofavorito** [comment]
BEFORE MERGE:
- [ ] check if `language-agnostic-definition.md` needs to be updated.

File: scripts/generate_all_protocols.py
Comment (marcofavorito) [inline scripts/generate_all_protocols.py:1]: TODO

File: docs/agent-vs-aea.md
Comment (marcofavorito) [reply docs/agent-vs-aea.md]: @DavidMinarsch UPDATE: please see ca534b0

File: aea/mail/base.proto
Comment (marcofavorito) [reply aea/mail/base.proto]: I decided to use `fetchai` for consistency with other parts of the tech stack (e.g. GitHub username, AEA package author username etc.); see later changes

File: aea/mail/base.proto
Comment (marcofavorito) [inline aea/mail/base.proto:13]: this was `google.protobuf.Any`; now it is bytes

File: aea/test_tools/generic.py
Comment (marcofavorito) [reply aea/test_tools/generic.py]: with the new Proto definition it seems 2 are enough, see changes

File: aea/connections/stub/connection.py
Comment (marcofavorito) [inline aea/connections/stub/connection.py:221]: not quite related to the PR

File: aea/connections/stub/connection.py
Comment (DavidMinarsch) [reply aea/connections/stub/connection.py:79]: We cannot merge before we have addressed #1413..not safe imo

File: aea/mail/base.proto
Comment (DavidMinarsch) [inline aea/mail/base.proto:3]: should be `package aea;` imo or maybe `aea.base` or something like that

File: aea/protocols/base.py
Comment (DavidMinarsch) [inline aea/protocols/base.py:291]: we should never add the message_id, dialogue_ref and target to the body then?

File: scripts/generate_all_protocols.py
Comment (DavidMinarsch) [reply scripts/generate_all_protocols.py:1]: still to do?

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]

File: aea/protocols/base.py
Comment (marcofavorito) [reply aea/protocols/base.py:291]: the problem is that they are in the body; they have been used above with `msg.message_id` etc., and we need to retrieve the rest of the body so we avoid to include the dialogue info fields twice

**Comment by marcofavorito** [comment]
https://github.com/fetchai/agents-aea/pull/1715#discussion_r492788990

@DavidMinarsch some changes have been lost due to merging after 441ec9a. I rewrote the history from there. Have a look now

File: aea/connections/stub/connection.py
Comment (lrahmani) [reply aea/connections/stub/connection.py:79]: The main limiting factor with the stub connection input file parsing is the need to support `echo`ing a serialized message in text format. I believe that providing an `aea` cli helper that constructs a message from command line arguments and serialize it to the stdout will free the stub connection to use any encoding/decoding format.

File: scripts/generate_all_protocols.py
Comment (marcofavorito) [reply scripts/generate_all_protocols.py:1]: No, check now.

File: aea/mail/base.proto
Comment (lrahmani) [reply aea/mail/base.proto:3]: @marcofavorito Why not `aea.fetchai` to be coherent with the rest of protos?

File: aea/mail/base.proto
Comment (lrahmani) [inline aea/mail/base.proto:18]: How useful is `google.protobuf.Struct` in comparison to a plain `string` field containing a JSON object serialization?

**Review by lrahmani — APPROVED** [review]
Looks good to me

File: aea/mail/base.proto
Comment (DavidMinarsch) [reply aea/mail/base.proto:3]: because `fetchai` refers to the author of the package/protocol. The base mail class has no author in that sense...

File: aea/protocols/base.py
Comment (DavidMinarsch) [reply aea/protocols/base.py:291]: That's correct

File: aea/connections/stub/connection.py
Comment (DavidMinarsch) [reply aea/connections/stub/connection.py:79]: We already have `aea interact` command. What's missing in the stub connection is specifying the length of the bytes string in the encoding so we can do away with the hack.

File: aea/protocols/generator/base.py
Comment (DavidMinarsch) [inline aea/protocols/generator/base.py:1761]: Good!

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, missing stub v2 fix and renaming of base proto.

File: fetchai_private_key.txt
Comment (DavidMinarsch) [inline fetchai_private_key.txt:1]: please remove

File: tests/test_docs/test_build_aea_programmatically/fetchai_private_key.txt
Comment (DavidMinarsch) [inline tests/test_docs/test_build_aea_programmatically/fetchai_private_key.txt:1]: please remove

**Review by DavidMinarsch — APPROVED** [review]
LGTM, please remove accidentally committed files before merging

**Pull request merged**
